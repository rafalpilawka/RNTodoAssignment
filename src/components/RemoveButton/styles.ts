import {StyleSheet} from 'react-native';
import normalize from 'react-native-normalize';

export const styles = StyleSheet.create({
  text: {color: 'white', fontSize: 16, fontWeight: 'bold'},
  button: {
    backgroundColor: 'red',
    borderRadius: normalize(50),
    width: normalize(30),
    height: normalize(30),
    justifyContent: 'center',
    alignItems: 'center',
  },
});
