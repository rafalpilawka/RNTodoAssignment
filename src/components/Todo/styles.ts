import {StyleSheet} from 'react-native';
import normalize from 'react-native-normalize';

export const styles = StyleSheet.create({
  todo: {
    flexDirection: 'row',
    height: normalize(80),
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomColor: 'black',
    borderBottomWidth: 1,
  },
  text: {fontSize: normalize(16), flex: 1, paddingHorizontal: normalize(20)},
});
