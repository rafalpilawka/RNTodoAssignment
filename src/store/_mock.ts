export const mockTodos = [
  {id: 'aaa', title: 'Add those todos', completed: false},
  {id: 'aaa2233', title: 'Delete second', completed: false},
  {id: 'bbnjhjhjhjh', title: 'Add third one', completed: true},
];

export const mockLists = [
  {name: 'List 1', id: '11242342144'},
  {name: 'list 2', id: '11242342122'},
  {name: 'list 3', id: '1124234211'},
];

export const mockStore = {
  index: mockLists,
  todos: mockTodos,
  loading: false,
  errors: undefined,
  idLoader: undefined,
};

export const mockInitialStore = {
  index: [],
  todos: [],
  loading: false,
  errors: undefined,
  idLoader: undefined,
};
