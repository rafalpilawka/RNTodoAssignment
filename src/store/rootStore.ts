import {all, call} from 'redux-saga/effects';
import {applyMiddleware, combineReducers, createStore, Reducer} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import todosReducer from 'src/store/todos/todos.reducer';
import {watchTodosSaga} from 'src/store/todos/todos.saga';

const sagaMiddleware = createSagaMiddleware();

function* rootSaga() {
  yield all([
    call(watchTodosSaga),
    //OTHER WATCHERS GOES HERE
  ]);
}

// const store: Reducer<RootState> = combineReducers<RootState>({
//   todos: todosReducer,
// });

const rootStore = (state: any, action: any) => {
  return todosReducer(state, action);
};

// export interface RootState {
//   todos: ReduxState;
// }

export type RootState = ReturnType<typeof todosReducer>;

export default createStore(
  rootStore,
  composeWithDevTools(applyMiddleware(sagaMiddleware)),
);
sagaMiddleware.run(rootSaga);
