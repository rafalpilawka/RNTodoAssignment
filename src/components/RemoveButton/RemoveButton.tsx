import React, {ReactElement} from 'react';
import {Pressable, Text} from 'react-native';
import {styles} from 'src/components/RemoveButton/styles';

export type PropsT = {
  onPress(): void;
};

export const RemoveButton = ({onPress}: PropsT): ReactElement => {
  return (
    <Pressable style={styles.button} onPress={onPress}>
      <Text style={styles.text}>x</Text>
    </Pressable>
  );
};
