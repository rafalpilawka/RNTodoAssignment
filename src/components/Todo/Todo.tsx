import React, {ReactElement} from 'react';
import {Pressable, Text} from 'react-native';
import {RemoveButton} from 'src/components/RemoveButton/RemoveButton';
import {TodoType} from 'src/store/todos/todos.types';
import CheckBox from '@react-native-community/checkbox';
import {styles} from 'src/components/Todo/styles';

type PropsT = {
  item: TodoType;
  removeHandler: (id: string) => void;
  toggleHandler: (id: string) => void;
};

export const Todo = ({
  removeHandler,
  toggleHandler,
  item: {id, title, completed},
}: PropsT): ReactElement => {
  const _removeTodo = () => {
    removeHandler(id);
  };

  const _toggleCheckBox = () => {
    toggleHandler(id);
  };

  return (
    <Pressable style={styles.todo} onLongPress={() => console.log('longPress')}>
      <CheckBox value={completed} onValueChange={_toggleCheckBox} />
      <Text style={styles.text}>{title}</Text>
      <RemoveButton onPress={_removeTodo} />
    </Pressable>
  );
};
