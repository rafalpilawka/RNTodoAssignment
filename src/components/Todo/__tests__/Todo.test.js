import React from 'react';
import 'react-native';
import {shallow} from 'enzyme';
import {Todo} from 'src/components/Todo/Todo';
import renderer from 'react-test-renderer';
import {mockTodos} from 'src/store/_mock';
import {RemoveButton} from 'src/components/RemoveButton/RemoveButton';
import CheckBox from '@react-native-community/checkbox';

const props = {
  item: mockTodos[1],
  removeHandler: jest.fn(),
  toggleHandler: jest.fn(),
};

describe('Todo item', () => {
  let component;
  beforeEach(() => {
    component = shallow(<Todo {...props} />);
    jest.clearAllMocks();
  });
  it('should match to snapshot', () => {
    const component = renderer.create(<Todo {...props} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should receive props', () => {
    const titleLabel = component.find({children: 'Delete second'});
    const completeStatus = component.props().children[0].props.value;
    expect(titleLabel).toHaveLength(1);
    expect(completeStatus).toBeFalsy();
  });

  it('should call removeHandler', () => {
    const removeButton = component.find(RemoveButton);
    removeButton.simulate('press');
    expect(props.removeHandler).toHaveBeenCalledTimes(1);
  });
  it('should call toggleHandler', () => {
    const toggleHandler = component.find(CheckBox);
    toggleHandler.simulate('valueChange');
    expect(props.toggleHandler).toHaveBeenCalledTimes(1);
  });
});
