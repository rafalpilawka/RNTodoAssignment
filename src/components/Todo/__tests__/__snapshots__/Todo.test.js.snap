// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo item should match to snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderBottomColor": "black",
      "borderBottomWidth": 1,
      "flexDirection": "row",
      "height": 160,
      "justifyContent": "space-between",
    }
  }
>
  <View
    pointerEvents="auto"
  >
    <RNCCheckbox
      forwardedRef={null}
      onValueChange={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "height": 32,
            "width": 32,
          },
          undefined,
        ]
      }
      value={false}
    />
  </View>
  <Text
    style={
      Object {
        "flex": 1,
        "fontSize": 32,
        "paddingHorizontal": 40,
      }
    }
  >
    Delete second
  </Text>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "borderRadius": 100,
        "height": 60,
        "justifyContent": "center",
        "width": 60,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      x
    </Text>
  </View>
</View>
`;
